### PR TITLE
ddl, expression: disable nondeterministic function calls for generated columns

### PR DESCRIPTION
**Pull request by spongedu**
### What problem does this PR solve? 
As is described in [CREATE TABLE and Generated Columns](https://dev.mysql.com/doc/refman/5.7/en/create-table-generated-columns.html) and [ALTER TABLE and Generated Columns](https://dev.mysql.com/doc/refman/5.7/en/alter-table-generated-columns.html), `generated` columns should not refer to any nondeterministic columns.

The Detail cases can be found in [MySQL test cases](https://github.com/mysql/mysql-server/blob/5.7/mysql-test/suite/gcol/inc/gcol_blocked_sql_funcs_main.inc )

### What is changed and how it works?
Add related check on ddl.

### Check List 

Tests 

 - Unit test
 - Integration test

Code changes

Side effects

Related changes

This pr depends on related error definition, which refers to another pr in [parser](https://github.com/pingcap/parser/pull/194).
After [parser related changes](https://github.com/pingcap/parser/pull/194) has been merged, the modification on `go.mod` and `go.sum` should be reverted.

### CODE REVIEW

**Comment by tiancaiamao** [comment]
LGTM

File: ddl/ddl.go [hunk 1]
```diff
@@ -109,6 +109,7 @@ var (
 	errTooManyFields            = terror.ClassDDL.New(codeTooManyFields, "Too many columns")
 	errInvalidSplitRegionRanges = terror.ClassDDL.New(codeInvalidRanges, "Failed to split region ranges")
 	errReorgPanic               = terror.ClassDDL.New(codeReorgWorkerPanic, "reorg worker panic.")
+	ErrGeneratedColumnFunctionIsNotAllowed               = terror.ClassDDL.New(codeErrGeneratedColumnFunctionIsNotAllowed, "Expression of generated column '%s' contains a disallowed function.")
```
Comment (zz-jason) [inline ddl/ddl.go]: we should add this error message to the parser repo?

File: executor/ddl_test.go [hunk 2]
```diff
@@ -593,3 +593,46 @@ func (s *testSuite3) TestSetDDLReorgBatchSize(c *C) {
...
+	c.Assert(err.Error(), Equals, ddl.ErrGeneratedColumnFunctionIsNotAllowed.GenWithStackByArgs("b").Error())
+
+	tk.MustExec("create table t1 (a bigint not null primary key auto_increment, b bigint, c bigint as (b + 1));")
+
+	_, err = tk.Exec("alter table t1 add column d varchar(1024) generated always as (database());")
+	c.Assert(err.Error(), Equals, ddl.ErrGeneratedColumnFunctionIsNotAllowed.GenWithStackByArgs("d").Error())
+
+	tk.MustExec("alter table t1 add column d bigint generated always as (b + 1); ")
+
+	_, err = tk.Exec("alter table t1 modify column d bigint generated always as (connection_id());")
```
Comment (zimulala) [inline executor/ddl_test.go]: I think we need to add `checkIllegalFn4GeneratedColumn` to `ModifyColumn/ChangeColumn`(`getModifiableColumnJob`). Otherwise this test will fail.

File: ddl/ddl.go (hunk 1 above)
Comment (spongedu) [reply ddl/ddl.go]: yes, done in https://github.com/pingcap/parser/pull/194

File: executor/ddl_test.go (hunk 2 above)
Comment (spongedu) [reply executor/ddl_test.go]: ok , Will fix this ~

Comment (spongedu) [reply executor/ddl_test.go]: Done~ @zimulala  PTAL

**Comment by zz-jason** [comment]
@spongedu Please fix lint:
```
linting

  ✘  https://revive.run/r#exported  exported var ErrGeneratedColumnFunctionIsNotAllowed should have comment or be unexported  

  ddl/ddl.go:113:2

✘ 1 problem (1 error, 0 warnings)
```

File: ddl/ddl_api.go [hunk 3]
```diff
@@ -2002,6 +2012,9 @@ func setDefaultAndComment(ctx sessionctx.Context, col *table.Column, options []*
 			col.Flag |= mysql.OnUpdateNowFlag
 			setOnUpdateNow = true
 		case ast.ColumnOptionGenerated:
+			if err := checkIllegalFn4GeneratedColumn(col.Name.L, opt.Expr); err != nil {
```
Comment (zimulala) [inline ddl/ddl_api.go]: Could we put this check to `checkModifyGeneratedColumn`? 
And I think it's best to unify the check for the generated column, otherwise, it's easy to forget. However, this may need to be handled in another PR

Comment (spongedu) [reply ddl/ddl_api.go]: Yes, it's better to put the checks for generated columns together. There's several other problems for generated columns, maybe we should make another PR to refactor this place later.

**Comment by spongedu** [comment]
@zz-jason @zimulala PTAL

File: ddl/ddl_api.go (hunk 3 above)
Comment (zimulala) [reply ddl/ddl_api.go]: In this PR, Could we put this check to `checkModifyGeneratedColumn`?

Comment (spongedu) [reply ddl/ddl_api.go]: @zimulala  I considered to put this check into `checkModifyGeneratedColumn` before, but changed my mind later because that means we need to go through the `Options` one more time. 
What do you think about this?~

**Comment by winkyao** [comment]
@spongedu Please fix conflicts and ci

**Comment by spongedu** [comment]
@winkyao  OK ~

File: ddl/ddl_api.go (hunk 3 above)
Comment (zimulala) [reply ddl/ddl_api.go]: It seems that the `ColumnOptionGenerated` option has only one per column and we only do `checkIllegalFn4GeneratedColumn` when the `ColumnOptionGenerated` option is selected. So we needn't
 to traverse the `options` in `checkModifyGeneratedColumn`?

File: ddl/ddl_api.go [hunk 4]
```diff
@@ -2205,6 +2218,14 @@ func (d *ddl) AlterColumn(ctx sessionctx.Context, ident ast.Ident, spec *ast.Alt
 		return errBadField.GenWithStackByArgs(colName, ident.Name)
 	}
 
+	for _, option := range specNewColumn.Options {
+		if option.Tp == ast.ColumnOptionGenerated {
+			if err := checkIllegalFn4GeneratedColumn(colName.L, option.Expr); err != nil {
```
[the remark this answers is not in the document]
Comment (spongedu) [reply ddl/ddl_api.go]: It seems that `ALTER TABLE ALTER COLUMN` do not modify generated columns' expressions , so there's no need for this check here

File: ddl/ddl_api.go (hunk 3 above)
Comment (spongedu) [reply ddl/ddl_api.go]: OK, I'll put `checkIllegalFn4GeneratedColumn` into `checkModifyGeneratedColumn` here

**Comment by spongedu** [comment]
@zimulala @winkyao @crazycs520 @zz-jason  PTAL :)
